### ISSUE TITLE
Tracking Issue for `#![feature(pointer_try_cast_aligned)]`

### ISSUE THREAD

**Issue by mathisbot** [rust-lang/rust#141221]
# Feature gate:  `#![feature(pointer_try_cast_aligned)]`

This is a tracking issue for adding the convenience method `try_cast_aligned` to `*const T`, `*mut T` and `NonNull<T>`.

```rust
impl<T: ?Sized> *const T {
    #[must_use]
    #[inline]
    pub fn try_cast_aligned<U: Sized>(self) -> Option<*const U>;
}

impl<T: ?Sized> *mut T {
    #[must_use]
    #[inline]
    pub fn try_cast_aligned<U: Sized>(self) -> Option<*mut U>;
}

impl<T: ?Sized> NonNull<T> {
    #[must_use]
    #[inline]
    pub fn try_cast_aligned<U: Sized>(self) -> Option<NonNull<U>>;
}
```

## Steps / History

- [X] API change proposal (ACP): https://github.com/rust-lang/libs-team/issues/588
- [ ] Implementation
- [ ] Final comment period (FCP)
- [ ] Stabilization PR

## Unresolved Questions

None yet.

**Comment by mathisbot** [comment]
## Stabilization report: `ptr::try_cast_aligned`

### Summary

This report proposes the stabilization of `try_cast_aligned` on pointers, which attempts to cast a pointer to a target type while checking alignment requirements.

```rust
impl<T: ?Sized> *const T {
    pub fn try_cast_aligned<U: Sized>(self) -> Option<*const U>;
}

impl<T: ?Sized> *mut T {
    pub fn try_cast_aligned<U: Sized>(self) -> Option<*mut U>;
}

impl<T: ?Sized> NonNull<T> {
    pub fn try_cast_aligned<U: Sized>(self) -> Option<NonNull<U>>;
}
```

The method returns `Some` if the pointer is properly aligned for `U`, and `None` otherwise.
It also ensures the checked alignment and the resulting pointer type are consistent.

### Motivation

Working with raw pointers often requires ensuring proper alignment before performing casts or de-referencing. Currently, users must write patterns such as:

```rust
assert!(ptr.cast::<U>().is_aligned());
let ptr = ptr.cast::<U>();
```

This approach has several drawbacks:

* it requires casting before checking alignment,
* it is still flagged by `clippy::cast_ptr_alignment`,
* it obscures the intent of the code.

### Alternatives

The original ACP proposed `ptr::is_aligned_for::<U>() -> bool` as the primary API.

This approach is closely related to the existing unstable API in [https://github.com/rust-lang/rust/issues/96284](https://github.com/rust-lang/rust/issues/96284) (`is_aligned_to`), which provides alignment checks independently of casting.

### Conclusion

The feature has been available on nightly for an extended period (more than one year) and while usage appears to be relatively limited, no negative feedback has been raised during this period.

I believe this feature is ready for stabilization.

cc @rust-lang/libs-api
